subscriber.py

```diff
@@ -0,0 +1,230 @@
+    logger.info(f"Creating permission for {log_group_arn}...")
+
+    lambda_client.add_permission(
+        FunctionName=lambda_arn,
+        StatementId=statement_id,
+        Action="lambda:InvokeFunction",
+        Principal=f"logs.amazonaws.com",
+        SourceArn=log_group_arn,
+    )
+
+
+def remove_permission(statement_id: str, lambda_arn: str):
+    lambda_client.remove_permission(
+        FunctionName=lambda_arn,
+        StatementId=statement_id,
+    )
+
+
+def create_subscription_filter(log_group_arn: str, lambda_arn: str):
+    log_group_name = log_group_arn.split(":")[-2]
+    logger.info(f"Creating subscription filter for {log_group_name}")
+
+    filter_name = "%s-axiom" % log_group_name
+
+    cloudwatch_logs_client.put_subscription_filter(
+        logGroupName=log_group_name,
+        filterName=filter_name,
+        filterPattern="",
+        destinationArn=lambda_arn,
+        distribution="ByLogStream",
+    )
+    logger.info(f"{log_group_name} subscription filter has been created successfully.")
+
+
+def lambda_handler(event: dict, context=None):
+    # handle deletion of the stack
+    if event["RequestType"] == "Delete":
+        cfnresponse.send(
+            event, context, cfnresponse.SUCCESS, {}, event["PhysicalResourceId"]
+        )
+        return
+
+    if (
+        axiom_cloudwatch_forwarder_lambda_arn is None
+        or axiom_cloudwatch_forwarder_lambda_arn == ""
+    ):
+        raise Exception("AXIOM_CLOUDWATCH_LAMBDA_FORWARDER_ARN is not set")
+
+    aws_account_id = context.invoked_function_arn.split(":")[4]
+    region = os.getenv("AWS_REGION")
+
+    log_group_names_list = (
+        log_group_names.split(",") if log_group_names is not None else []
+    )
+    log_groups = build_groups_list(
+        get_log_groups(), log_group_names_list, log_group_pattern, log_group_prefix
+    )
+
+    # report number of log groups found
+    logger.info(f"Found {len(log_groups)} log groups that matches the criteria.")
+
+    report = {
+        "log_groups_count": len(log_groups),
+        "matched_log_groups": [],
+        "added_groups": [],
+        "added_groups_count": 0,
+        "errors": {},
+    }
+    for group in log_groups:
+        # skip the Forwarder lambda log group to avoid circular logging
+        if group["name"].startswith("/aws/axiom/"):
+            continue
+
+        # create invoke permission for lambda
+        cleaned_name = "-".join(group["name"].split("/")[3:])
+        statement_id = f"invoke-permission-for-{cleaned_name}"
+
+        report["matched_log_groups"].append(group["name"])
+        report["errors"][group["name"]] = []
+
+        # if the log group already have a subscription filter, skip it
+        try:
+            response = cloudwatch_logs_client.describe_subscription_filters(
+                logGroupName=group["name"]
+            )
+            # TODO: improve the Subscription filters check
+            if len(response["subscriptionFilters"]) > 0:
```

Comment:
I think a better check would be the destinationArn == functionArn?

Reply:
This would check if the log group is subscribed to the same Forwarder, which is not the case. Here, we need to check if the log group has subscription for another/previous Forwarder and avoid overriding it. Not sure what is a good solution here.

@dominicchapman This is part of the mechanism of the **Subscriber** that tries to help the users overcome the limit size of permissions. Currently, when a Subscriber is checking which groups to add subscription on, it checks if there is any subscription filters on it, if there is already, it will skip it. This might not be what the users want since they might have subscriptions that send to other places, or maybe be they intend to have that log group forwarded to 2 different datasets, which they won't be able to do here. What's your thoughts?